A/B - Low-friction participation test.

gtrufitt opened the pull request:
## What does this change?

This adds the functionality for an A/B test for low-friction participation, initially star-ratings.

## What is the value of this and can you measure success?

We believe that low-friction participation like star ratings and generic 'Did you like this article' type questions at the end of articles will increase participation across the Guardian.

For each of the low-friction tests, **success looks like** at least double the amount of clicks on the low-friction features as comments on the same articles.

## Does this affect other platforms - Amp, Apps, etc?

This is web-only currently but a test on Apps would be interesting.

## Screenshots

![initial](https://cloud.githubusercontent.com/assets/638051/15330528/22487e68-1c55-11e6-9633-3e28d246dfd6.gif)

## Request for comment

@GHaberis, @NathanielBennett  - If you could review the code that'd be awesome, thanks!

cc/ @crifmulholland @guardian/participation 

- [x] X-Browser
- [x] Pull in page title to description for this specific test
- [x] Improve the accessibility and keyboard nav
- [x] Add listener for action

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/low-friction-participation.js:
```diff
@@ -0,0 +1,261 @@
+    mediator
+) {
+
+    var currentState = {
```
We don't really use the single var style any more...

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/low-friction-participation.js:
```diff
@@ -0,0 +1,261 @@
+
+    // Mark-up Building
+
+    var createButtons = function (state) {
```
Why declare these functions as variables?

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/low-friction-participation.js:
```diff
@@ -0,0 +1,261 @@
+
+        setUserVote();
+
+        mediator.emit('modules:participation:clicked', currentState.selectedItem);
```
What's the second param used for?

gtrufitt replied:
Right now, nothing, so I'll remove

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/low-friction-participation.js:
```diff
@@ -0,0 +1,261 @@
+define([
```
Some stylistic comments (thought I'd pull these out of the body of the diff)

- we don't really use single-var declarations any more
- not sure if we need to define functions as vars

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/low-friction-participation.js:
```diff
@@ -0,0 +1,261 @@
+        updateState({
+            confirming: true,
+            selectedItem: $(event.currentTarget).data().itemId,
+            initialRender: false
```
It's not a huge issue for experiment code, but I'm not sure I'd make these functions poke around internal state vars like `initialRender`. But - for an A/B test it hardly matters.

sndrs commented on static/src/stylesheets/module/experiments/_ab-low-friction-participation.scss:
```diff
@@ -0,0 +1,48 @@
+.participation-low-fric--button__is-active:focus .inline-star,
+.participation-low-fric--button__is-active:active .inline-star,
+.participation-low-fric--button__is-highlighted .inline-star {
+    fill: #ffce4b;
```
👀 ?

jbreckmckye commented on static/src/javascripts/projects/common/views/experiments/participation/low-friction-buttons.html:
```diff
@@ -0,0 +1,6 @@
+<<%=elType%>
```
What do you think of having a single template, with the different permutations in if-blocks? I wonder if it might tweaking and rapidly iterating a little quicker. I guess it could also grow unwieldy quite quickly, too, so YMMV.

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/tests/participation-star-ratings.js:
```diff
@@ -36,7 +38,7 @@ define([
         this.author = 'Gareth Trufitt - Participation';
         this.description = 'Initial user segmentation to ensure statistical significance';
-        this.audience = 1;
+        this.audience = 0;
```
So, is this only an opt-in test, for now?

gtrufitt replied:
Yup.

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/tests/participation-star-ratings.js:
```diff
@@ -45,7 +47,10 @@ define([
+            return config.page.section === 'film' &&
+                config.page.toneIds === 'tone/reviews' &&
+                config.page.commentable &&
+                Object.create; // Filters out IE8.
```
You could also check `isEnhanced`, if that's more appropriate.

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/tests/participation-star-ratings.js:
```diff
@@ -45,7 +47,10 @@ define([
-            return config.page.section === 'film' && config.page.toneIds === 'tone/reviews' && config.page.commentable;
+            return config.page.section === 'film' &&
+                config.page.toneIds === 'tone/reviews' &&
+                config.page.commentable &&
```
Why commentable?

gidsg replied:
I think because the participation baseline is comparison with commentable articles, so only with this condition do we have a valid comparison point.

jbreckmckye replied:
Sensible.

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/tests/participation-star-ratings.js:
```diff
@@ -66,7 +71,33 @@ define([
                 id: 'star-rating',
-                test: function () {}
+                test: function () {
+                    if (Object.create) {
```
You already tested this?

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/tests/participation-star-ratings.js:
```diff
@@ -66,7 +71,33 @@ define([
-                test: function () {}
+                test: function () {
+                    if (Object.create) {
+                        var starRatings = Object.create(lowFrictionParticipation),
```
Why not just use `new`?

GHaberis replied:
LowFrictionParticipation isn't a constructor so I don't think you could use the new operator here

jbreckmckye replied:
Yeah, but you could make it one.

jbreckmckye commented on static/src/javascripts/projects/common/modules/experiments/tests/participation-star-ratings.js:
```diff
@@ -66,7 +71,33 @@ define([
+                            description = '';
+
+                        if (config.page.headline && config.page.author) {
+                            description = 'Is your rating the same as ' + config.page.author + ' on "' + config.page.headline + '"?';
```
Strictly, shouldn't `config.page.author` be in the possessive? e.g. "Is your rating the same as Mark Kermode's on _Son of Saul_?"

You'd have to handle names ending in _s_, though:

```javascript```
function possessive(name) {
  var lastChar = name.substr(-1);
  var postfix = (lastChar === 's') ? "'" : "'s";
  return name + postfix;
}

gtrufitt commented:
Hi @jbreckmckye - Thanks for your review!

I've now removed the single var declarations and no longer varring my functions, I moved a lot of view logic into the views and got rid of a couple of the templates. I left the button looping in the JS as actually the logic in the view is not that nice, I'll maybe review that more in the future though. I also added your possessive.

For the test, I am going to leave the object.create stuff but I have started refactoring in a different branch but also wanted to pull in the mini-redux pattern so will do that if this graduates out of a test.

Could you take another look and let me know if there's anything else?

jbreckmckye commented:
Looks great, Gareth - and it'll be interesting to see how users respond.

👍

gtrufitt commented:
Thanks Jimmy!
